### DISCUSSION TITLE
Support for pyvista?

### DISCUSSION THREAD

**Discussion by fdinhSG** [zauberzeug/nicegui#745] [Q&A] [answered]
### Question

Hey! I'm trying to build an app where I need to view some data using pyvista. Is this supported by niceGUI?

I've seen that 3d viewing uses three.js, I was wondering if I was able to somehow include pyvista too.

Looking at the pyvista docs I saw it is possible to [include it in flask](https://docs.pyvista.org/extras/flask.html), so I guess it's feasible with any framework, I just don't know if niceGUI has the flexibility to allow it too.

Thanks!

**Comment by falkoschindler** [comment, marked as answer]
Interesting idea, @fdinhSG! That is absolutely possible with NiceGUI.

As in the Flask example, you can embed the generated HTML code with an iframe:

```py
import pyvista
from nicegui import app, ui

pl = pyvista.Plotter(window_size=(800, 600))
pl.add_mesh(pyvista.Sphere())
pl.export_html('/tmp/sphere.html', 'panel')

app.add_static_files('/meshes', '/tmp')
ui.html('<iframe src="/meshes/sphere.html" width="800" height="600"></iframe>')

ui.run()
```

Strangely, I had to use the "panel" backend, since the default "pythreejs" backend did not work.

**Reply by fdinhSG** [reply]
Thank you! I'm sorry but I wasn't able to replicate your script on my pc, what's in the `sphere.html` file you're referring to?

Thanks again and sorry, I'm kinda new to programming and I'm trying to figure out lots of new things.

**Reply by falkoschindler** [reply]
No worries. The sphere.html file should be generated when starting the app. Maybe you have to adjust the paths `'/tmp/sphere.html'` and `'/tmp'` if you're on Windows; I'm not sure.

**Reply by fdinhSG** [reply]
Mmmmh I think there might be an issue with Windows folders indeed, plus other weird things that might have to do with my configuration. The `export_html` works without issues without the `panel` option but with it an error message appears: `RuntimeError: Nothing to screenshot - call .show first or use the off_screen argument`

That error message disappeared with

```python
import pyvista, os
from nicegui import app, ui

pl = pyvista.Plotter(window_size=(800, 600))
pl.add_mesh(pyvista.Sphere())

tmp = os.path.join(os.getcwd(),'tmp')
pl.export_html(os.path.join(tmp,'sphere.html'), backend='panel')

app.add_static_files('/meshes', tmp)
ui.html('<iframe src="/meshes/sphere.html" width="800" height="600"></iframe>')

ui.run()
```

But, strangely enough, line 8 is giving me an [`unknown file extension .html` issue](https://user-images.githubusercontent.com/128691970/231973019-e371e445-b056-4688-8ca6-3e835ea8c941.png). I'm wondering if it's still a Windows issue or if it's something with the way `export_html` works.

Anyways, I think I've taken enough of your time and I believe this is more of a pyvista issue than an nicegui one. Thanks for your support! I'll see if I can solve the `export_html` issue and I'll try to implement it in nicegui once that's done.

**Comment by gsal** [comment]
Hi, 

I know this is closed, but I was doing my due diligence before posting and found this discussion; I too am interested in using nicegui and then pyvista as a friendlier interface to do some vtk stuff.

I looked at trame and it seemed a bit convoluted and too vtk-centric; for any other widgets typical of a dashboard application, one needs to visit Vuetify, look the widget docs in javascript, and 'translate' to python based on some rules in the trame doc itself...not cool. 

And, so, I rather learn nicegui as the all a round frontend for my applications and then bring in pyvista when needing vtk stuff. 

Does anybody know how to properly include a pyvista widget to visualize a vtk file? and possibly have whatever   

[ Sorry, I don't know much but the code above seems a bit of a hack? ]

**Reply by fdinhSG** [reply]
Sorry! I was away for the holidays.

I don't know if it's the answer you're looking for, but I ended up going with trame. I agree with you that vuetify is not as intuitive as niceGUI, but at the end of the day it's the de-facto supported solution and I preferred going with it instead of trying to bind two separate systems together. pyvista and trame are both developed by kitware so they speak the same language, in a sense.

trame also has a discussion forum on github and the main dev is super helpful for any issue you have.

**Comment by gsal** [comment]
Oops, I guess I have no way of re-opening this discussion; will create a new one, then.
